core/src/main/java/ai/timefold/solver/core/impl/bavet/common/AbstractStaticDataNode.java

```diff
@@ -0,0 +1,159 @@
+        var state = tuple.state;
+        if (state.isDirty()) {
+            if (state == TupleState.DYING || state == TupleState.ABORTING) {
+                throw new IllegalStateException("The fact (%s) was retracted, so it cannot update."
+                        .formatted(a));
+            }
+            // CREATING or UPDATING is ignored; it's already in the queue.
+        } else {
+            propagationQueue.update(tuple);
+        }
+    }
+
+    @Override
+    public final void retract(Object a) {
+        if (!tupleMap.containsKey(a)) {
+            return;
+        }
+        invalidateCache();
+        tupleMap.remove(a);
+        retractFromNodeNetwork(a);
+        recalculateTuples();
+    }
+
+    private void insertNew(Tuple_ tuple) {
+        var state = tuple.state;
+        if (state != TupleState.CREATING) {
+            propagationQueue.insert(tuple);
+        }
+    }
+
+    private void retractExisting(Tuple_ tuple) {
+        var state = tuple.state;
+        if (state.isDirty()) {
+            if (state == TupleState.DYING || state == TupleState.ABORTING) {
+                // We already retracted this tuple from another list, so we
+                // don't need to do anything
+                return;
+            }
+            propagationQueue.retract(tuple, state == TupleState.CREATING ? TupleState.ABORTING : TupleState.DYING);
+        } else {
+            propagationQueue.retract(tuple, TupleState.DYING);
+        }
+    }
+
+    private void insertIntoNodeNetwork(Object toInsert) {
+        nodeNetwork.getTupleSourceRootNodes(toInsert.getClass())
+                .forEach(node -> ((AbstractForEachUniNode) node).insert(toInsert));
+    }
+
+    private void retractFromNodeNetwork(Object toRetract) {
+        nodeNetwork.getTupleSourceRootNodes(toRetract.getClass())
+                .forEach(node -> ((AbstractForEachUniNode) node).retract(toRetract));
+    }
+
+    private void invalidateCache() {
+        tupleMap.values().stream().flatMap(List::stream).forEach(this::retractExisting);
+        recordingTupleNode.getTupleRecorder().reset();
+    }
+
+    private void recalculateTuples() {
+        // Settle all the inserts/retracts that happen
+        nodeNetwork.settle();
+        var recorder = recordingTupleNode.getTupleRecorder();
+        for (var mappedTupleEntry : tupleMap.entrySet()) {
+            mappedTupleEntry.getValue().clear();
+            var invalidated = mappedTupleEntry.getKey();
+            recorder.recordingInto(mappedTupleEntry.getValue(), this::remapTuple, () -> {
+                // Do an update on the object and settle the network; this will update precisely the
+                // tuples mapped to this node, which will then be recorded
+                nodeNetwork.getTupleSourceRootNodes(invalidated.getClass())
+                        .forEach(node -> ((AbstractForEachUniNode) node).update(invalidated));
+                nodeNetwork.settle();
```

Comment:
Why would you settle on every update? Is that necessary for anything?
I'd rather settle after all operations had been queued.